dask/dataframe/io/io.py

```diff
@@ -750,5 +754,145 @@ def sorted_division_locations(seq, npartitions=None, chunksize=None):
     return values, positions
 
 
+@insert_meta_param_description
+def from_map(
+    func,
+    inputs,
+    meta=None,
+    divisions=None,
+    label=None,
+    token=None,
+    enforce_metadata=True,
+    **kwargs,
+):
+    """Create a DataFrame collection from a custom function map
+
+    Parameters
+    ----------
+    func : function
+        Function used to create each partition.
+    inputs : Iterable
+        Iterable object of arguments to pass to ``func``. The number of
+        elements in ``inputs`` will correspond to the number of partitions
+        in the output collection (only one element will be passed to
+        ``func`` for each partition).
+    $META
+    divisions : tuple, str, optional
+        Partition boundaries along the index.
+        For tuple, see https://docs.dask.org/en/latest/dataframe-design.html#partitions
+        For string 'sorted' will compute the delayed values to find index
+        values.  Assumes that the indexes are mutually sorted.
+        If None, then won't use index information
+    label : str, optional
+        String to use as the function-name label in the output
+        collection-key names.
+    token : str, optional
+        String to use as the "token" in the output collection-key names.
+    enforce_metadata : bool, default True
+        Whether to enforce at runtime that the structure of the DataFrame
+        produced by ``func`` actually matches the structure of ``meta``.
+        This will rename and reorder columns for each partition,
+        and will raise an error if this doesn't work or types don't match.
+    **kwargs:
+        Key-word arguments to be passed to ``func`` for every output partition.
+
+    Examples
+    --------
+    >>> import pandas as pd
+    >>> import dask.dataframe as dd
+    >>> func = lambda x, size=0: pd.Series([x] * size)
+    >>> inputs = ["A", "B"]
+    >>> dd.from_map(func, inputs, size=2).compute()
+    0    A
+    1    A
+    0    B
+    1    B
+    dtype: object
+
+    See Also
+    --------
+    dask.dataframe.from_delayed
+    dask.layers.DataFrameIOLayer
+    """
+
+    if not callable(func):
+        raise ValueError("`func` argument must be `callable`")
+    if not isinstance(inputs, Iterable):
+        raise ValueError(f"`inputs` must be Iterable, got {type(inputs)}")
+    inputs = list(inputs)  # Always convert to list (for now)
+    if not len(inputs):
+        raise ValueError("`inputs` must have a non-zero length")
```

Comment:
Why this restriction? I could imagine some useful functions that produce random testing data and don't rely on any arguments.